# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
data_children_pivot = data.pivot_table(index=['children'],  values='debt', aggfunc = ['count','sum', 'mean'])

print(data_children_pivot)

          count   sum      mean
           debt  debt      debt
children                       
0         14091  1063  0.075438
1          4808   444  0.092346
2          2052   194  0.094542
3           330    27  0.081818
4            41     4  0.097561
5             9     0  0.000000


In [23]:
# КОД РЕВЬЮЕРА

display(data.head(2))
display(data.head(3))
data.head(4)

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   

     family_status  family_status_id gender income_type  debt  total_income  \
0  женат / замужем                 0      F   сотрудник     0        253875   
1  женат / замужем                 0      F   сотрудник     0        112080   

                   purpose total_income_category          purpose_category  
0            покупка жилья                     B  операции с недвижимостью  
1  приобретение автомобиля                     C    операции с автомобилем

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   среднее             1   

     family_status  family_status_id gender income_type  debt  total_income  \
0  женат / замужем                 0      F   сотрудник     0        253875   
1  женат / замужем                 0      F   сотрудник     0        112080   
2  женат / замужем                 0      M   сотрудник     0        145885   

                   purpose total_income_category          purpose_category  
0            покупка жилья                     B  операции с недвижимостью  
1  приобретение автомобиля                     C    операции с автомобилем  
2            покупка жилья                     C  операции с недвижимостью

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   среднее             1   
3         3    4124.747207         32   среднее             1   

     family_status  family_status_id gender income_type  debt  total_income  \
0  женат / замужем                 0      F   сотрудник     0        253875   
1  женат / замужем                 0      F   сотрудник     0        112080   
2  женат / замужем                 0      M   сотрудник     0        145885   
3  женат / замужем                 0      M   сотрудник     0        267628   

                      purpose total_income_category          purpose_category  
0               покупка жилья                     B  операции с недвижимостью  
1     приобретение автомобиля                     C    операции с автомобилем  
2               покупка жилья                     C  операции с недвижимостью  
3  дополнительное образование                     B     получение образования

**Вывод:** 

При иследовании зависимсти между количеством детей и возвратом кредита в срок выявили следующие данные. 
Самый большой процент невозврата кредита 9,76% с семьях с 4-мя детьми (таких семей всего 41, и на 41 семью приходится 4 невозрата). 
На втором и третьем месте семьи с 2-мя и с 1-им ребенком: 9,45% и 9,23% соответственно, процент возврата незначительно различается, а вот количество семей с 1-им ребенком в 2,3 раза превышает количество семей с 2-мя детьми (так же как и случаев невозврата).
На 4-ом месте семьи с 3-мя детьми - невозврат составляет 8,18%.
На 5-ом месте семьи без детей (таких семей в  2 раза больше, чем семей с детьми) - невозврат составляет 7,54%.
и на 6-ом месте семьи с 5-ю детьми - там нет случаев невозврата.

In [24]:
data_children_pivot = data.pivot_table(index = ['children'],  values = 'debt', aggfunc = {'count', 'sum', 'mean'})
data_children_pivot['percent'] = data_children_pivot['mean']*100
data_children_pivot = data_children_pivot.sort_values(by='percent', ascending=False)
display(data_children_pivot)

count      mean     sum   percent
children                                     
4            41.0  0.097561     4.0  9.756098
2          2052.0  0.094542   194.0  9.454191
1          4808.0  0.092346   444.0  9.234609
3           330.0  0.081818    27.0  8.181818
0         14091.0  0.075438  1063.0  7.543822
5             9.0  0.000000     0.0  0.000000

При иследовании зависимсти между количеством детей и возвратом кредита в срок выявили следующие данные. <br>
<ul>
 <li> Самый большой процент невозврата кредита 9,76% с семьях с 4-мя детьми (таких семей всего 41, и на 41 семью приходится 4 невозрата). </li>
 <li> На втором и третьем месте семьи с 2-мя и с 1-им ребенком: 9,45% и 9,23% соответственно, процент возврата незначительно различается, а вот количество семей с 1-им ребенком в 2,3 раза превышает количество семей с 2-мя детьми (так же как и случаев невозврата).  </li>
 <li> На 4-ом месте семьи с 3-мя детьми - невозврат составляет 8,18%. <li>На 5-ом месте семьи без детей (таких семей в 2 раза больше, чем семей с детьми) - невозврат составляет 7,54%.  </li>
 </ul> <br> Cемьи с 5-ю детьми не имеют случаев невозврата. Данную категорию в анализе учитывать неккоректно, так как случаев невозврата не выявлено, категория совсем малочисленная. И, к тому же, у семей с 4-мя детьми доля просрочек самая большая, было бы ошибочным сделать вывод о том, что семьи с 5-ю детьми являются более надежными заемщики, чем с 4-мя.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [25]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
data_family_pivot = data.pivot_table(index=['family_status'],  values='debt', aggfunc = {'count','sum', 'mean'})
data_family_pivot['percent'] = data_family_pivot['mean']*100
data_family_pivot = data_family_pivot.sort_values(by='percent', ascending=False)
display(data_family_pivot)

count      mean    sum   percent
family_status                                            
Не женат / не замужем   2796.0  0.097639  273.0  9.763948
гражданский брак        4134.0  0.093130  385.0  9.313014
женат / замужем        12261.0  0.075606  927.0  7.560558
в разводе               1189.0  0.070648   84.0  7.064760
вдовец / вдова           951.0  0.066246   63.0  6.624606

**Вывод:** 

При иследовании зависимсти между семейным положение и возвратом кредита в срок выявили следующие данные.<br>
<ul> 
<li>Самый высокий процент невозврата кредита у заемщиков со статусом "не женат / не замужем" и "гражданский брак" - 9,76% и 9,31% соответственно.</li>
<li>Самый низкий процент невозврата в семьях со статусом "вдовец / вдова" - 6,62%.</li>
<li>Средние провенты имеют семьи со статусом "женат / замужем" и "в разводе" - 7,56% и 7,06% соответственно.</li><br>
Возможно, на эти данные влияет и возраст заемщиков, так как "вдовец / вдова" - это чаще всего люди в возрасте и имеющие взрослых детей, а "не женат / не замужем" и "гражданский брак" - молодые люди.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [26]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
data_income_pivot = data.pivot_table(index=['total_income_category'],  values='debt', aggfunc = {'count','sum', 'mean'})
data_income_pivot['percent'] = data_income_pivot['mean']*100
data_income_pivot = data_income_pivot.sort_values(by='percent', ascending=False)
display(data_income_pivot)

count      mean     sum   percent
total_income_category                                     
E                         22.0  0.090909     2.0  9.090909
C                      15921.0  0.084982  1353.0  8.498210
A                         25.0  0.080000     2.0  8.000000
B                       5014.0  0.070602   354.0  7.060231
D                        349.0  0.060172    21.0  6.017192

**Вывод:** 

При иследовании зависимсти между уровнем дохода и возвратом кредита в срок выявили следующие данные.<br>
<ul> 
<li>Самый высокий процент невозврата кредита в семьях с доходом категории "Е" - 9,09%, это группа с самым низким уровнем дохода (до 30000). </li>
<li>На 2-ом месте по невозвратам группа "С" - 8,50% (группа с доходом от 50000 до 200000).</li>
<li>На 3-ем месте по невозвратам группа "А" - 8,00% (группа с доходом от 1 000 000).</li>
<li>На 4-ом месте по невозвратам группа "В" - 7,06% (группа с доходом от 200 000 до 1 000 000).</li>
<li>Самый низкий процент невозврата в семьях с доходом категории "D" - 6,02%, это группа с  достатком от 30000 до 50000.</li><br> Если брать количество случаев в каждой категории, то мы видим, что больше всего заемщиков с доходом от 50 000 до 20000 и заемщиков с доходом от 200 000 до 1 000 000. Данные вызывают сомнения, так как средняя зарплата по России более низкая. Следовательно, анализировать заемщиков именно по этим показателям некорректно, слишком разное колличество случаев в категориях. Возможно, нужна более узкая разбивка на категории или исследовать вместе с доходом дополнительный параметр(например, регион).

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [27]:
data_purpose_pivot = data.pivot_table(index=['purpose_category'],  values='debt', aggfunc = {'count','sum', 'mean'})
data_purpose_pivot['percent'] = data_purpose_pivot['mean']*100
data_purpose_pivot = data_purpose_pivot.sort_values(by='percent', ascending=False)
display(data_purpose_pivot)

count      mean    sum   percent
purpose_category                                            
операции с автомобилем     4279.0  0.093480  400.0  9.347978
получение образования      3988.0  0.092528  369.0  9.252758
проведение свадьбы         2313.0  0.079118  183.0  7.911803
операции с недвижимостью  10751.0  0.072551  780.0  7.255139

**Вывод:** 

При иследовании зависимсти между уровнем дохода и возвратом кредита в срок выявили следующие данные.<br>
<ul> 
<li>Самый высокий процент невозврата кредита имеют кредиты с целью "операции с автомобилем" и "получение образования" - 9,35% и 9,25% соответственно.</li>
<li>Самый низкий процент невозврата кредита имеют кредиты с целью "операции с недвижимостью" и "проведение свадьбы " - 7,73% и 7,91% соответственно.</li><br>
Но количество кредитов на "операции с недвижимостью" практически равны количеству кредитов на остальные нужды.


#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 

Пропуски встречались в столбцах "days_employed" и "total_income" ("общий стаж в днях" и "доход ежемесячный"). Причем, количество пропусков в обоих столбцах одинаково и равно 2174. Возможно, что незаполнение одного столбца ведет к незаполнению другого.<br>
Пропуски, скорее всего, носят случайный характер, так как непредоставление сведений о доходах при больших суммах кредита (на приобретение недвижимости, например) приведут к неодобрению кредита. Следовательно, это не умышленное неуказание данных.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* 

В столбцах, в которых необходимо было заполнить пропуски, очень большой числовой разброс. Следовательно, если брать среднее арифметическое по данным, то анализ данных будет не совсем корректным. Медианное же значение лучше всего характеризует выборки и больше подходит для заполнения пропусков.

### Шаг 4: общий вывод.

<p>По проведенному анализу можно слелать следующий вывод. </p> <p>Следующие категории граждан имеют высокий процент риска невозврата кредита: <ul> <li>заемщики с низким уровнем дохода (ниже 30 000) 9,09%,</li> <li>либо заемщики с 4-мя детьми 9,96%,</li> <li>либо заемщик со статусом "не женат / не замужем" 9,76% и "гражданский брак" 9,31%,</li> <li>либо цель кредита  "операции с автомобилем" 9,35% и "получение образования" 9,25%.</li> <p> Риск невозврата повышается, если совпадет несколько из этих категорий. </p> <p>Наиболее надежными заемщиками являются следующие категории:</p> <ul> <li>заемщики с целью кредита "операции с недвижимостью" и "проведение свадьбы " - 7,73% и 7,91% соответственно</li> <li>с семейным статусом "женат / замужем" и "в разводе" - 7,56% и 7,06% соответственно, либо "вдовец/вдова" 6,62% (но количество случаев в этой категории мало)</li> <li>семьи с 3-мя детьми 8,18%</li> <li> и семьи с доходом категории "D" - 6,02%, это группа с достатком от 30 000 до 50 000.
